### ISSUE TITLE
Sub-optimal code generation for extern "C" fn type calls

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang/rust#64090]
See https://rust.godbolt.org/z/9UvEyu

```rust
![feature(unwind_attributes)]

extern "C" { 
    #[unwind(allow)] fn foo(); 
    // fn bar();
    static bar: extern "C" fn(); 
    static mut BAR: i32;
}

struct Foo;
impl Drop for Foo {
    fn drop(&mut self) {
        unsafe { BAR = 42; }
    }
}

pub unsafe fn unwind() { 
    let x = Foo;
    foo(); 
    std::mem::forget(x);
}
pub unsafe fn nounwind() { 
    let x = Foo;
    bar(); 
    std::mem::forget(x);
}
```

When the function `extern "C" { fn bar(); }` is called, the `nounwind` function is compiled to:

```asm
example::nounwind:
        jmpq    *bar@GOTPCREL(%rip)
```

However, when the function type `static bar: extern "C" fn();` is called, this sub-optimal machine code is emitted:

```asm
example::nounwind:
        pushq   %rbx
        movq    bar@GOTPCREL(%rip), %rax
        callq   *(%rax)
        popq    %rbx
        retq
        movq    %rax, %rbx
        callq   core::ptr::real_drop_in_place
        movq    %rbx, %rdi
        callq   _Unwind_Resume@PLT
        ud2
```

**Comment by RalfJung** [comment]
I don't think this is a bug. At least if we accept https://github.com/rust-lang/rfcs/pull/2753, until we can track the unwinding ABI in fn ptr types, we cannot and should not assume that functions can't unwind based only on their call ABI.

Before changing our codegen for fn ptrs, IMO we should accept an RFC that basically says the opposite of https://github.com/rust-lang/rfcs/pull/2753.

**Comment by nagisa** [comment]
FWIW LLVM has no way of tracking "nounwind" for function pointers and since `bar` could be anything, LLVM cannot assume the pointer is nounwind.

**Comment by gnzlbg** [comment]
@nagisa we can apply attributes to calls through function pointers, e.g., see how clang does it: https://rust.godbolt.org/z/YMv6NJ

@RalfJung That discussion belongs to the RFC PR.

**Comment by nagisa** [comment]
> @nagisa we can apply attributes to calls through function pointers, e.g., see how clang does it: https://rust.godbolt.org/z/YMv6NJ

The code generated by `clang` applies no attributes to its arguments in LLVM-IR, All that happens is clang carefully selecting whether to generate `call` or `invoke` during lowering to LLVM, which is different. Though it is true that Rust could do the same, but this then touches heavily upon how certain RFCs will go.

**Comment by Centril** [comment]
> certain RFCs will go.

@nagisa specifically which ones?

**Comment by gnzlbg** [comment]
@nagisa See https://godbolt.org/z/O67kh1, we currently optimize

```rust
extern "C" { fn foo(); } 
static FOO: unsafe extern "C" fn() = foo;
pub unsafe fn bar() { FOO() } 
```

to essentially this

```rust
extern "C" { fn foo(); } 
pub unsafe fn bar() { foo() } // bar is nounwind because foo is nounwind
```

This optimization is sound if the `extern "C"` ABI of the function pointer is the same as the `extern "C"` ABI of the function declaration, which in this case, it does never unwind, and this allows us to remove the static, remove the landing pads in bar, and make bar nounwind. 

[Declaring a function with the wrong ABI is probably not UB, calling it is.](https://github.com/rust-lang/unsafe-code-guidelines/issues/198) Here, we never call the declaration, only the function pointer. If it were ok for the function pointer to unwind, then AFAICT at least two of the optimizations above (removing the landing pads and making bar nounwind) are unsound. 

So AFAICT we are already optimizing Rust code under the assumption that the `extern "C"` ABI of function pointers matches that of function declarations and definitions. We are just not as clever as clang when it comes to choosing `call` over `invoke` when we know that an ABI does not unwind.

**Comment by comex** [comment]
Even if we had a separate `extern "C+unwind"` or similar, we’d probably translate calls to function pointers of that ABI to the same IR that we currently generate for `extern "C"`.  Thus, if you declared a function as `extern "C"`, created a pointer to it, casted that to an `extern "C+unwind"` function pointer, and called the new pointer, you’d run into the same optimization.

As such, either:
- The optimization is incorrect; or
- If a function declaration has the wrong ABI, it’s UB not only to call the function directly, but also to call a function pointer derived from that declaration, even if the pointer is casted to the correct type first.

Either way, it seems like an orthogonal issue to whether or not `extern "C"` function pointers should be implicitly unwind-capable.

**Comment by RalfJung** [comment]
> So AFAICT we are already optimizing Rust code under the assumption that the extern "C" ABI of function pointers matches that of function declarations and definitions.

We are optimizing based on the assumption that calling a function through a function pointer behaves the same as if calling the function directly.
*If* we can figure out the function the pointer points to, and *if* that function is `nounwind`, that lets us assume that the fn ptr call is also nounwind. And that's what we see happening here.

**Comment by RalfJung** [comment]
> That discussion belongs to the RFC PR.

Well, right now we are (in our handling of fn ptrs) future compatible with `extern "C"` functions existing that *can* unwind. Given those RFCs, now is a bad time to remove that bit of future compatibility.

> specifically which ones?

https://github.com/rust-lang/rfcs/pull/2753, I imagine.

**Comment by RalfJung** [comment]
> Even if we had a separate extern "C+unwind" or similar, we’d probably translate calls to function pointers of that ABI to the same IR that we currently generate for extern "C". Thus, if you declared a function as extern "C", created a pointer to it, casted that to an extern "C+unwind" function pointer, and called the new pointer, you’d run into the same optimization.

I have no idea what you mean by this. Calling an `extern "C+unwind"` fn ptr would generate IR that can handle unwinding out of that call. What is the problem?

Unwinding from an `extern "C"` fn is UB (in a world with an `"C+unwind"` ABI), and the fact that we call an `extern "C"` fn via an `extern "C+unwind"` ptr changes nothing. In general, when the caller and callee signature differ, *both* signatures must be "correct" for the data that is actually used. For example, if you call a `fn(bool)` by transmuting the ptr to `fn(u8)` and passing in `0x04`, that's UB because the callee expects a `bool`. But also, if you have a `fn(u8)` and transmute it to `fn(bool)` and then pass in `0x04`, you produced a `0x04` at type `bool` so you caused UB. So for unwinding, if *either* the caller or callee signature says "no unwinding", then it is UB if unwinding happens.

**Comment by gnzlbg** [comment]
> I have no idea what you mean by this. Calling an extern "C+unwind" fn ptr would generate IR that can handle unwinding out of that call. What is the problem?

If you do this:

```rust
extern "C" {
   fn foo(); // Declared to never unwind
}
static FOO: extern "C" unsafe fn() = foo;
static BAR: extern "C+unwind" unsafe fn() = foo as _; // this cast is sound
unsafe fn bar() { BAR() } // BOOM: bar is optimized as nounwind
```

We will optimize `bar` under the assumption that `BAR()` and `foo()` do not unwind because `foo` is declared as `nounwind`. 

So either this optimization is unsound _or_ it is UB to call a function declared as not-unwinding via a function pointer that supports unwinding. 

IMO, this is just another issue of:

```rust
mod bad { extern "C" { fn foo(); } } // never unwinds
extern "C+unwind" { fn foo(); }  // BOOM: becomes nounwind because of bad::foo
```

**Comment by RalfJung** [comment]
> So either this optimization is unsound or it is UB to call a function declared as not-unwinding via a function pointer that supports unwinding if that function actually unwinds.

That certainly is UB, as I said above.

It is UB for a function to unwind if at least one of the caller and callee signature says this may not unwind. In your case, the callee signature declared here
```rust
extern "C" { fn foo(); } // Declared to never unwind
```
says "cannot unwind", and hence this may not unwind, no matter the signature used by the caller.

**Comment by comex** [comment]
But that's not the callee signature; it's only a signature we're using on the caller's end to declare the symbol.  I'm assuming that the callee's actual implementation (which might be in another crate, or written in a different language altogether) is properly declared as unwinding.

**Comment by RalfJung** [comment]
I would say it is the callee signature; if we have both declarations *and* a definition of the same symbol floating around and they are not the same then we are looking at https://github.com/rust-lang/unsafe-code-guidelines/issues/198 which is an orthogonal issue.
